### PR TITLE
Proposed fix to add setters tests

### PR DESCRIPTION
**Pull request by lemire**
This is a PR on add-setters-tests. It improves the performance somewhat, but we still have a large regression compared to the main branch. Prior to merging any of this into the main branch, we should understand the cause of the large regression.

```
This PR:
BasicBench_AdaURL       7985 ns         7985 ns        83977 time/byte=9.88189ns time/url=725.87ns url/s=1.37766M/s
BasicBench_AdaURL       7934 ns         7933 ns        87971 time/byte=9.81829ns time/url=721.198ns url/s=1.38658M/s
BasicBench_AdaURL       7904 ns         7903 ns        88095 time/byte=9.7807ns time/url=718.437ns url/s=1.39191M/s
BasicBench_AdaURL       7947 ns         7946 ns        88526 time/byte=9.83446ns time/url=722.386ns url/s=1.3843M/s

the original setters PR
BasicBench_AdaURL       8432 ns         8432 ns        83380 time/byte=10.4356ns time/url=766.543ns url/s=1.30456M/s
BasicBench_AdaURL       8425 ns         8424 ns        82965 time/byte=10.4263ns time/url=765.856ns url/s=1.30573M/s
BasicBench_AdaURL       8433 ns         8431 ns        83155 time/byte=10.435ns time/url=766.498ns url/s=1.30464M/s
BasicBench_AdaURL       8383 ns         8382 ns        83157 time/byte=10.3741ns time/url=762.023ns url/s=1.3123M/s

Main branch:
BasicBench_AdaURL       7242 ns         7241 ns        96974 time/byte=8.96132ns time/url=658.25ns url/s=1.51918M/s
BasicBench_AdaURL       7195 ns         7194 ns        97116 time/byte=8.90391ns time/url=654.033ns url/s=1.52898M/s
BasicBench_AdaURL       7220 ns         7218 ns        97100 time/byte=8.93345ns time/url=656.203ns url/s=1.52392M/s
BasicBench_AdaURL       7268 ns         7267 ns        96063 time/byte=8.99431ns time/url=660.673ns url/s=1.51361M/s
```

### CODE REVIEW

File: include/ada/parser.h [hunk 1]
```diff
@@ -15,9 +15,9 @@ namespace ada::parser {
   std::optional<ada::url_host> parse_ipv6(std::string_view input);
   std::optional<ada::url_host> parse_host(std::string_view input, bool is_not_special, bool input_is_ascii);
 
-  url parse_url(std::string user_input,
+  url parse_url(std::string_view user_input,
```
Comment (anonrig) [inline include/ada/parser.h:18]: I think we also need to update all user facing functions to use string view

File: src/implementation.cpp [hunk 2]
```diff
@@ -5,37 +5,54 @@
...
-    return ada::parser::parse_url(std::string(input));
+    // TODO std::move(base_url) might be unwise. Check.
+    return ada::parser::parse_url(input, std::move(base_url), encoding, std::nullopt, state);
   }
 
   /*
    * The protocol setter steps are to basic URL parse the given value, followed by U+003A (:),
    * with this’s URL as url and scheme start state as state override.
    *
+   * TODO: This should probably a method in the struct ada::url.
```
Comment (anonrig) [inline src/implementation.cpp:29]: I agree. I had to stop in the middle of my pull request. Appreciate if you can move this.

File: src/implementation.cpp [hunk 3]
```diff
@@ -5,37 +5,54 @@
...
+  void set_scheme(ada::url& base, std::string input, ada::encoding_type encoding) noexcept {
     if (!input.empty()) {
       input.append(":");
     }
 
-    auto result = ada::parser::parse_url(input, std::nullopt, encoding, base, SCHEME_START);
+    /**
+     * TODO: This needs to be reengineered. The next line calls
+     * a large function just to later update the scheme. We should
+     * specialize and just call what is needed: a scheme computation.
```
Comment (anonrig) [inline src/implementation.cpp:41]: I agree. Eventually we need to do this for all states, and reuse these functions in the main function

**Comment by lemire** [comment]
Merging directly to [add-setters-tests](https://github.com/ada-url/ada/tree/add-setters-tests) where I will be doing more work.

**Pull request merged**